### PR TITLE
Add table azure_nat_gateway closes #525

### LINKED ISSUE #525: Add table azure_nat_gateway
**References**
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/nat_gateway

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/azure_nat_gateway []

PRETEST: tests/azure_nat_gateway

TEST: tests/azure_nat_gateway
Running terraform
data.azurerm_client_config.current: Reading...
azurerm_resource_group.named_test_resource: Refreshing state... [id=/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest93685]
data.azurerm_client_config.current: Read complete after 0s [id=Y2xpZW50Q29uZmlncy9jbGllbnRJZD0wNGIwNzc5NS04ZGRiLTQ2MWEtYmJlZS0wMmY5ZTFiZjdiNDY7b2JqZWN0SWQ9MDZmZDQ2YjAtYTg2Ny00OWExLWE0ZjEtZjc3Njg0NjVjYWJhO3N1YnNjcmlwdGlvbklkPWQ0NmQ3NDE2LWY5NWYtNDc3MS1iYmI1LTUyOWQ0Yzc2NjU5Yzt0ZW5hbnRJZD1jZGZmZDcwOC03ZGEwLTRjZWEtYWJlYi0wYTRjMzM0ZDdmNjQ=]
data.null_data_source.resource: Reading...
data.null_data_source.resource: Read complete after 0s [id=static]
azurerm_nat_gateway.named_test_resource: Refreshing state... [id=/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest93685/providers/Microsoft.Network/natGateways/turbottest93685]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the
last "terraform apply" which may have affected this plan:

  # azurerm_nat_gateway.named_test_resource has been deleted
  - resource "azurerm_nat_gateway" "named_test_resource" {
      - id                      = "/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest93685/providers/Microsoft.Network/natGateways/turbottest93685" -> null
        name                    = "turbottest93685"
      - resource_guid           = "d79aa045-5c0e-4f7a-8800-887d05f97856" -> null
        tags                    = {
            "name" = "turbottest93685"
        }
        # (5 unchanged attributes hidden)
    }

  # azurerm_resource_group.named_test_resource has been deleted
  - resource "azurerm_resource_group" "named_test_resource" {
        id       = "/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest93685"
      - location = "eastus" -> null
      - name     = "turbottest93685" -> null
    }


Unless you have made equivalent changes to your configuration, or ignored the
relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # azurerm_nat_gateway.named_test_resource will be created
  + resource "azurerm_nat_gateway" "named_test_resource" {
      + id                      = (known after apply)
      + idle_timeout_in_minutes = 10
      + location                = "eastus"
      + name                    = "turbottest60556"
      + resource_group_name     = "turbottest60556"
      + resource_guid           = (known after apply)
      + sku_name                = "Standard"
      + tags                    = {
          + "name" = "turbottest60556"
        }
      + zones                   = [
          + "1",
        ]
    }

  # azurerm_resource_group.named_test_resource will be created
  + resource "azurerm_resource_group" "named_test_resource" {
      + id       = (known after apply)
      + location = "eastus"
      + name     = "turbottest60556"
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  ~ resource_aka       = "azure:///subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest93685/providers/Microsoft.Network/natGateways/turbottest93685" -> (known after apply)
  ~ resource_aka_lower = "azure:///subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourcegroups/turbottest93685/providers/microsoft.network/natgateways/turbottest93685" -> (known after apply)
  ~ resource_guid      = "d79aa045-5c0e-4f7a-8800-887d05f97856" -> (known after apply)
  ~ resource_id        = "/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest93685/providers/Microsoft.Network/natGateways/turbottest93685" -> (known after apply)
  ~ resource_name      = "turbottest93685" -> "turbottest60556"
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 4s [id=/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556]
azurerm_nat_gateway.named_test_resource: Creating...
azurerm_nat_gateway.named_test_resource: Still creating... [10s elapsed]
azurerm_nat_gateway.named_test_resource: Creation complete after 19s [id=/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556/providers/Microsoft.Network/natGateways/turbottest60556]

Warning: Deprecated

  with data.null_data_source.resource,
  on variables.tf line 29, in data "null_data_source" "resource":
  29: data "null_data_source" "resource" {

The null_data_source was historically used to construct intermediate values
to re-use elsewhere in configuration, the same can now be achieved using
locals

(and one more similar warning elsewhere)

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = "azure:///subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556/providers/Microsoft.Network/natGateways/turbottest60556"
resource_aka_lower = "azure:///subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourcegroups/turbottest60556/providers/microsoft.network/natgateways/turbottest60556"
resource_guid = "fcf3de79-f243-4d5d-aa3f-ee0b8c91b072"
resource_id = "/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556/providers/Microsoft.Network/natGateways/turbottest60556"
resource_name = "turbottest60556"
subscription_id = "u85j7416-f95f-4371-bbf5-529v7c763847"

Running SQL query: test-get-query.sql
[
  {
    "id": "/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556/providers/Microsoft.Network/natGateways/turbottest60556",
    "name": "turbottest60556"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556/providers/Microsoft.Network/natGateways/turbottest60556",
    "name": "turbottest60556"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbottest60556/providers/Microsoft.Network/natGateways/turbottest60556",
      "azure:///subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourcegroups/turbottest60556/providers/microsoft.network/natgateways/turbottest60556"
    ],
    "name": "turbottest60556",
    "tags": {
      "name": "turbottest60556"
    },
    "title": "turbottest60556"
  }
]
✔ PASSED

POSTTEST: tests/azure_nat_gateway

TEARDOWN: tests/azure_nat_gateway

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select
  name,
  id,
  provisioning_state,
  sku_name,
  type
from
  azure_nat_gateway;
+--------+-----------------------------------------------------------------------------------------------------------------------------+--------------------+----------+-------------------------------+
| name   | id                                                                                                                          | provisioning_state | sku_name | type                          |
+--------+-----------------------------------------------------------------------------------------------------------------------------+--------------------+----------+-------------------------------+
| test53 | /subscriptions/u85j7416-f95f-4371-bbf5-529v7c763847/resourceGroups/turbot_rg/providers/Microsoft.Network/natGateways/test53 | Succeeded          | Standard | Microsoft.Network/natGateways |
+--------+-----------------------------------------------------------------------------------------------------------------------------+--------------------+----------+-------------------------------+

> select
	n.name,
	i.ip_address as ip_address,
  i.ip_configuration_id as ip_configuration_id,
  i.public_ip_address_version as public_ip_address_version,
  i.public_ip_allocation_method as public_ip_allocation_method
from
	azure_nat_gateway as n,
  azure_public_ip as i,
	jsonb_array_elements(n.public_ip_addresses) as ip
where
  ip ->> 'id' = i.id;
+--------+---------------+---------------------+---------------------------+-----------------------------+
| name   | ip_address    | ip_configuration_id | public_ip_address_version | public_ip_allocation_method |
+--------+---------------+---------------------+---------------------------+-----------------------------+
| test53 | 20.62.205.122 | <null>              | IPv4                      | Static                      |
+--------+---------------+---------------------+---------------------------+-----------------------------+

> select
	n.name as name,
	s.name as subnet_name,
	s.virtual_network_name as virtual_network_name
from
	azure_nat_gateway as n,
  azure_subnet as s,
	jsonb_array_elements(n.subnets) as sb
where
	sb ->> 'id' = s.id;
+--------+-------------+----------------------+
| name   | subnet_name | virtual_network_name |
+--------+-------------+----------------------+
| test53 | test12      | test-steampipe-vn    |
| test53 | default     | test-steampipe-vn    |
+--------+-------------+----------------------+
```
</details>

### CODE REVIEW

File: azure-test/tests/azure_nat_gateway/variables.tf [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+variable "azure_environment" {
+  type        = string
+  default     = "public"
+  description = "Azure environment used for the test."
+}
+
+variable "azure_subscription" {
+  type        = string
+  default     = "d46d7416-f95f-4771-bbb5-529d4c76659c"
```
Comment (bigdatasourav) [inline azure-test/tests/azure_nat_gateway/variables.tf:16]: Please put a dummy subscription ID here

File: azure-test/tests/azure_nat_gateway/variables.tf [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  inputs = {
+    scope = "azure:///subscriptions/${data.azurerm_client_config.current.subscription_id}"
+  }
+}
+
+resource "azurerm_resource_group" "named_test_resource" {
+  name     = var.resource_name
+  location = "East US"
+}
+
```
Comment (bigdatasourav) [inline azure-test/tests/azure_nat_gateway/variables.tf:39]: Remove the commented code

File: azure/table_azure_nat_gateway.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+	"github.com/turbot/steampipe-plugin-sdk/v4/plugin"
+)
+
+//// TABLE DEFINITION ////
+
+func tableAzureNatGateway(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "azure_nat_gateway",
+		Description: "Azure NAT Gateways",
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:18]:
```suggestion
		Description: "Azure NAT Gateway",
```

File: azure/table_azure_nat_gateway.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+			},
+			{
+				Name:        "id",
+				Description: "Contains ID to identify a nat gateway uniquely.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromGo(),
+			},
+			{
+				Name:        "etag",
+				Description: "An unique read-only string that changes whenever the resource is updated",
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:43]:
```suggestion
				Description: "An unique read-only string that changes whenever the resource is updated.",
```

File: azure/table_azure_nat_gateway.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+		}),
+	}
+}
+
+//// FETCH FUNCTIONS ////
+
+func listNatGateways(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+	session, err := GetNewSession(ctx, d, "MANAGEMENT")
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:140]: Add a logger here for the error

File: azure/table_azure_nat_gateway.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	}
+	subscriptionID := session.SubscriptionID
+
+	networkClient := network.NewNatGatewaysClient(subscriptionID)
+	networkClient.Authorizer = session.Authorizer
+
+	result, err := networkClient.ListAll(ctx)
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:149]: Add a logger here for the API error

File: azure/table_azure_nat_gateway.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+			}
+		}
+	}
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS ////
+
+func getNatGateway(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getNatGateway")
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:182]:
```suggestion
```

File: azure/table_azure_nat_gateway.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+	name := d.KeyColumnQuals["name"].GetStringValue()
+	resourceGroup := d.KeyColumnQuals["resource_group"].GetStringValue()
+
+	session, err := GetNewSession(ctx, d, "MANAGEMENT")
+	if err != nil {
+		return nil, err
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:189]: Add a logger here for the session error

File: docs/tables/azure_nat_gateway.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  sku_name,
+  type
+from
+  azure_nat_gateway;
+```
+
+### Public IP address info
+
+```sql
+select
```
Comment (bigdatasourav) [inline docs/tables/azure_nat_gateway.md:23]: Format is not correct for this query

File: azure/table_azure_nat_gateway.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+			{
+				Name:        "etag",
+				Description: "An unique read-only string that changes whenever the resource is updated",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "idle_timeout_in_minutes",
+				Description: "The idle timeout of the nat gateway.",
+				Type:        proto.ColumnType_INT,
+				Transform:   transform.FromP(extractNatGatewayProperties, "idleTimeoutInMinutes"),
```
Comment (bigdatasourav) [inline azure/table_azure_nat_gateway.go:50]: Can we use `transform.FromField("NatGatewayPropertiesFormat.IdleTimeoutInMinutes")` here and the same for the other columns where ever applicable to fetch the properties?

File: docs/tables/azure_nat_gateway.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  name,
+  id,
+  provisioning_state,
+  sku_name,
+  type
+from
+  azure_nat_gateway;
+```
+
+### Public IP address info
```
Comment (bigdatasourav) [inline docs/tables/azure_nat_gateway.md:20]:
```suggestion
### List public IP address details for each nat gateway
```

File: docs/tables/azure_nat_gateway.md [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+from
+	azure_nat_gateway as n,
+  azure_public_ip as i,
+	jsonb_array_elements(n.public_ip_addresses) as ip
+where
+  ip ->> 'id' = i.id;
+```
+
+
+### List subnet details associated with nat gatways
```
Comment (bigdatasourav) [inline docs/tables/azure_nat_gateway.md:38]:
```suggestion
### List subnet details associated with each nat gateway
```

File: azure-test/tests/azure_nat_gateway/variables.tf (hunk 2 above)
Comment (ParthaI) [reply azure-test/tests/azure_nat_gateway/variables.tf:39]: Removed

File: azure/table_azure_nat_gateway.go (hunk 3 above)
Comment (ParthaI) [reply azure/table_azure_nat_gateway.go:18]: Updated

File: azure/table_azure_nat_gateway.go (hunk 4 above)
Comment (ParthaI) [reply azure/table_azure_nat_gateway.go:43]: Updated

File: azure/table_azure_nat_gateway.go (hunk 10 above)
Comment (ParthaI) [reply azure/table_azure_nat_gateway.go:50]: we can not use `transform.FromField` due to inconsistency in API behavior we are not getting the inner-level property properly so we must need to extract it.

**Review by bigdatasourav — APPROVED** [review]
LGTM

**Pull request merged**